support/signal.c

```diff
@@ -166,6 +167,90 @@ keep_trying (int r)
 	return r == -1 && errno == EINTR;
 }
 
+// This tiny ad-hoc read/write lock is needed because of the very specific
+// synchronization needed between default_handler and teardown_pipes:
+// - Many default_handlers can be running at once
+// - External forces already ensure only one teardown_pipes runs at once
+// - If a teardown_pipes interrupts a default_handler, it must block
```

Comment:
I may be very naive, but *how* can `teardown_pipes` interrupt `default_handler` execution? What does "interrupt" mean in this context?

I can imagine `teardown_pipes()` *executing concurrently* with `default_handler()` on two separate threads (and two separate CPUs), but that's not "interrupt[ing execution]"...